File: api/models/dish.go
package models

import "errors"

type Course int

const (
	Starter Course = iota // 0
	Main                  // 1
	Dessert               // 2
	Side                  // 3
	Extra                 // 4
)

// Course to string map for getting the name of a course
var Courses = map[Course]string{
	Starter: "Starter",
	Main:    "Main",
	Dessert: "Dessert",
	Side:    "Side",
	Extra:   "Extra",
}

// Get the string value of a course
func (c Course) String() (string, error) {
	str, v := Courses[c]
	if !v {
		return "", errors.New("error, failed to find course ")
	}

	return str, nil
}

type Spice int

const (
	Mild     Spice = iota // 0
	Medium                // 1
	Hot                   // 2
	ExtraHot              // 3
)

// Spice to string map for getting the name of a spice level
var Spices = map[Spice]string{
	Mild:     "Mild",
	Medium:   "Medium",
	Hot:      "Hot",
	ExtraHot: "Extra Hot",
}

// Get a string value of the Spice
func (s Spice) String() (string, error) {
	str, v := Spices[s]
	if !v {
		return "", errors.New("error, failed to find course ")
	}

	return str, nil
}

// Dish item, inherits from the MenuItem struct
type Dish struct {
	Model
	ItemRefer  uint     `json:"itemRefer"`
	Item       MenuItem `gorm:"foreignKey:ItemRefer" json:"item"`
	Course     Course   `gorm:"not null;type:integer; check:(Course >= 0) and (Course <= 4)" json:"course"`
	Spice      Spice    `gorm:"not null;type:integer; check:(Spice >= 0) and (Spice <= 3)" json:"spice"`
	Vegetarian bool     `gorm:"not null;default:false" json:"vegetarian"`
}

File: web/lib/types/dish.ts
import { Item } from "./item";

export enum Spice {
  Mild = 0,
  Medium,
  Hot,
  ExtraHot,
}

const spiceMap: Map<Spice, String> = new Map([
  [Spice.Mild, "mild"],
  [Spice.Medium, "medium"],
  [Spice.Hot, "hot"],
  [Spice.ExtraHot, "extra hot"],
]);

export function getSpiceString(spice: Spice): string {
  const spiceString = spiceMap.get(spice);
  if (spiceString === undefined) {
    throw new Error("Invalid spice value");
  }

  return spiceString.toString();
}

export enum Course {
  Starter = 0,
  Main,
  Dessert,
  Side,
  Extra,
}

const courseMap: Map<Course, String> = new Map([
  [Course.Starter, "starter"],
  [Course.Main, "main"],
  [Course.Dessert, "dessert"],
  [Course.Side, "side"],
  [Course.Extra, "extra"],
]);

export function getCourseString(course: Course): string {
  const courseString = courseMap.get(course);
  if (courseString === undefined) {
    throw new Error("Invalid spice value");
  }

  return courseString.toString();
}

export type Dish = {
  id: number;
  item: Item;
  course: Course;
  spice: Spice;
  vegetarian: boolean;
};

export const getBlankDish = (): Dish => {
  return {
    id: 0,
    item: {
      id: 0,
      name: "Dish Name",
      description: "Dish Description",
      cost: 0,
      calories: 0,
      allergens: [""],
      available: false,
      image: "http://localhost:8080/images/dishes/blank.jpg",
    },
    course: 0,
    spice: 0,
    vegetarian: false,
  };
};
